cpp/src/arrow/dataset/partition.cc

```diff
@@ -18,92 +18,195 @@
-  std::vector<UnconvertedKey> keys(min);
-  for (int i = 0; i < min; i++) {
-    keys[i].name = schema_->field(i)->name();
-    keys[i].value = std::move(segments[i]);
+  Result<ExpressionPtr> Parse(const std::string& segment, int i) const override {
+    return scalar(true);
   }
+};
 
-  return ConvertPartitionKeys(keys, *schema_);
+PartitionSchemePtr PartitionScheme::Default() {
+  return std::make_shared<DefaultPartitionScheme>();
 }
 
-std::vector<UnconvertedKey> HivePartitionScheme::GetUnconvertedKeys(
-    const std::string& path) const {
-  auto segments = fs::internal::SplitAbstractPath(path);
+PartitionSchemeDiscovery::PartitionSchemeDiscovery() : schema_(::arrow::schema({})) {}
 
-  std::vector<UnconvertedKey> keys;
-  for (const auto& segment : segments) {
-    std::smatch matches;
-    static std::regex hive_style("^([^=]+)=(.*)$");
-    if (std::regex_match(segment, matches, hive_style) && matches.size() == 3) {
-      keys.push_back({matches[1].str(), matches[2].str()});
+Result<ExpressionPtr> SegmentDictionaryPartitionScheme::Parse(const std::string& segment,
+                                                              int i) const {
+  if (static_cast<size_t>(i) < dictionaries_.size()) {
+    auto it = dictionaries_[i].find(segment);
+    if (it != dictionaries_[i].end()) {
+      return it->second;
     }
   }
-  return keys;
+
+  return scalar(true);
+}
+
+Result<ExpressionPtr> PartitionKeysScheme::ConvertKey(const Key& key,
+                                                      const Schema& schema) {
+  auto field = schema.GetFieldByName(key.name);
+  if (field == nullptr) {
+    return scalar(true);
+  }
+
+  std::shared_ptr<Scalar> converted;
+  RETURN_NOT_OK(Scalar::Parse(field->type(), key.value, &converted));
+  return equal(field_ref(field->name()), scalar(converted));
+}
+
+Result<ExpressionPtr> PartitionKeysScheme::Parse(const std::string& segment,
+                                                 int i) const {
+  if (auto key = ParseKey(segment, i)) {
+    return ConvertKey(*key, *schema_);
+  }
+
+  return scalar(true);
+}
+
+util::optional<PartitionKeysScheme::Key> SchemaPartitionScheme::ParseKey(
+    const std::string& segment, int i) const {
+  if (i >= schema_->num_fields()) {
+    return util::nullopt;
+  }
+
+  return Key{schema_->field(i)->name(), segment};
+}
+
+inline bool AllIntegral(const std::vector<std::string>& reprs) {
+  return std::all_of(reprs.begin(), reprs.end(), [](string_view repr) {
+    // TODO(bkietz) use ParseUnsigned or so
+    return repr.find_first_not_of("0123456789") == string_view::npos;
+  });
+}
+
+inline std::shared_ptr<Schema> InferSchema(
+    const std::map<std::string, std::vector<std::string>>& name_to_values) {
+  std::vector<std::shared_ptr<Field>> fields(name_to_values.size());
+
+  size_t field_index = 0;
+  for (const auto& name_values : name_to_values) {
+    auto type = AllIntegral(name_values.second) ? int32() : utf8();
```

Comment:
Do you expect partition directories on timestamp seconds? The only use case I can see for seconds is as a segment in a deep ETL-style scheme like `/{year}/{month}/{day}/{hour}/{second}`, where it would only go from `0-3600`. Partitioning just on seconds `/{seconds_since_epoch}` would be a very flat and non performant scheme; do we really want to pessimize in order to support that? From the same logic follows my decision not to attempt to parse timestamps or floats.